src/Security/InheritedPermissions.php

```diff
@@ -0,0 +1,632 @@
+<?php
+
+namespace SilverStripe\Security;
+
+use InvalidArgumentException;
+use SilverStripe\Core\Injector\Injectable;
+use SilverStripe\ORM\DataList;
+use SilverStripe\ORM\DataObject;
+use SilverStripe\ORM\Hierarchy\Hierarchy;
+use SilverStripe\Versioned\Versioned;
+
+/**
+ * Calculates batch permissions for nested objects for:
+ *  - canView: Supports 'Anyone' type
+ *  - canEdit
+ *  - canDelete: Includes special logic for ensuring parent objects can only be deleted if their children can
+ *    be deleted also.
+ */
+class InheritedPermissions
+{
+    use Injectable;
+
+    const DELETE = 'delete';
+    const VIEW = 'view';
+    const EDIT = 'edit';
+
+    /**
+     * Anyone canView permission
+     */
+    const ANYONE = 'Anyone';
+
+    /**
+     * Restrict to logged in users
+     */
+    const LOGGED_IN_USERS = 'LoggedInUsers';
+
+    /**
+     * Restrict to specific groups
+     */
+    const ONLY_THESE_USERS = 'OnlyTheseUsers';
+
+    /**
+     * Inherit from parent
+     */
+    const INHERIT = 'Inherit';
+
+    /**
+     * Class name
+     *
+     * @var string
+     */
+    protected $baseClass = null;
+
+    /**
+     * Object for evaluating top level permissions designed as "Inherit"
+     *
+     * @var RootPermissions
+     */
+    protected $rootPermissions = null;
+
+    /**
+     * Global permissions required to edit.
+     * If empty no global permissions are required
+     *
+     * @var array
+     */
+    protected $globalEditPermissions = [];
+
+    /**
+     * Cache of permissions
+     *
+     * @var array
+     */
+    protected $cachePermissions = [];
+
+    /**
+     * @param RootPermissions $callback
+     * @return $this
+     */
+    public function setRootPermissions(RootPermissions $callback)
+    {
+        $this->rootPermissions = $callback;
+        return $this;
+    }
+
+    /**
+     * Global permissions required to edit
+     *
+     * @param array $permissions
+     * @return $this
+     */
+    public function setGlobalEditPermissions($permissions)
+    {
+        $this->globalEditPermissions = $permissions;
+        return $this;
+    }
+
+    /**
+     * @return array
+     */
+    public function getGlobalEditPermissions()
+    {
+        return $this->globalEditPermissions;
+    }
+
+    /**
+     * Get root permissions handler, or null if no handler
+     *
+     * @return RootPermissions|null
+     */
+    public function getRootPermissions()
+    {
+        return $this->rootPermissions;
+    }
+
+    /**
+     * Get base class
+     *
+     * @return string
+     */
+    public function getBaseClass()
+    {
+        return $this->baseClass;
+    }
+
+    /**
+     * Set base class
+     *
+     * @param string $baseClass
+     * @return $this
+     */
+    public function setBaseClass($baseClass)
```

Comment:
Shouldn't this be a constructor argument? The implementation will fail without this being set, right?